base/inference.jl

```diff
@@ -2607,21 +2607,12 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
     if !(isa(typea,Type) || isa(typea,TypeVar)) || !(isa(typeb,Type) || isa(typeb,TypeVar))
         return Any
     end
-    if (typea <: Tuple) && (typeb <: Tuple)
-        if isa(typea, DataType) && isa(typeb, DataType) && length(typea.parameters) == length(typeb.parameters) && !isvatuple(typea) && !isvatuple(typeb)
-            return typejoin(typea, typeb)
-        end
-        if isa(typea, Union) || isa(typeb, Union) || (isa(typea,DataType) && length(typea.parameters)>3) ||
-            (isa(typeb,DataType) && length(typeb.parameters)>3)
-            # widen tuples faster (see #6704), but not too much, to make sure we can infer
-            # e.g. (t::Union{Tuple{Bool},Tuple{Bool,Int}})[1]
-            return Tuple
-        end
-    end
     u = Union{typea, typeb}
-    if unionlen(u) > MAX_TYPEUNION_LEN || type_too_complex(u, MAX_TYPE_DEPTH)
-        # don't let type unions get too big
-        # TODO: something smarter, like a common supertype
+    if unionlen(u) > MAX_TYPEUNION_LEN
+        u = typejoin(typea, typeb)
```

Comment:
No, that cannot be the right interpretation of "monotonicity" in this context, there are too many cases where `tmerge` violates that. Is it sufficient that the sequence of `x`s obtained by repeatedly doing `x = tmerge(x, ...)` is monotonic? Then obviously `x >: tmerge(x, ...)` suffices, but also using `typejoin` should be kosher then. @vtjnash, you brought up monotonicity, could you clarify what you meant and what potential problems with `typejoin` you fear exactly?